rust/processor/src/db_writer.rs

```diff
@@ -0,0 +1,325 @@
+/***
+   The goal of this module is to move db writes to their own executor pool
+   This allows us to have a separate pool of threads for db writes
+   In the future, we'd like to use a rayon pool for the core processing, and isolate async code
+*/
+
+use crate::{
+    data_with_query::DataWithQuery,
+    utils::{
+        counters::{
+            DB_EXECUTION_RETRIES_COUNT, DB_EXECUTOR_CHANNEL_CHUNK_INSERT_COUNT,
+            DB_EXECUTOR_CHANNEL_SIZE, DB_INSERTION_ROWS_COUNT, DB_NULL_BYTE_INSERT_COUNT,
+            DB_SINGLE_BATCH_EXECUTION_TIME_IN_SECS,
+        },
+        database::{get_chunks, get_config_table_chunk_size, PgDbPool},
+        util::remove_null_bytes,
+    },
+    worker::PROCESSOR_SERVICE_TYPE,
+};
+use ahash::{AHashMap, AHashSet};
+use diesel::{
+    query_builder::QueryFragment,
+    result::{DatabaseErrorKind, Error},
+    QueryResult,
+};
+use kanal::{AsyncReceiver, AsyncSender};
+use tokio::task::JoinHandle;
+use tracing::info;
+
+pub trait DbCleanable {
+    fn clean(&mut self);
+}
+
+impl<Item> DbCleanable for Vec<Item>
+where
+    Item: field_count::FieldCount
+        + serde::Serialize
+        + for<'de> serde::Deserialize<'de>
+        + Send
+        + Clone,
+{
+    fn clean(&mut self) {
+        *self = self.iter().map(remove_null_bytes).collect()
+    }
+}
+
+#[allow(clippy::len_without_is_empty)]
+pub trait DbRowCountable {
+    fn len(&self) -> usize;
+}
+
+impl<Item> DbRowCountable for Vec<Item>
+where
+    Item: Send,
+{
+    fn len(&self) -> usize {
+        self.len()
+    }
+}
+
+#[async_trait::async_trait]
+/// The `DbExecutable` trait is for types that can be executed in the database
+pub trait DbExecutable: DbCleanable + DbRowCountable + Send {
+    async fn execute_query(&self, db_pool: PgDbPool) -> QueryResult<usize>;
+
+    // TODO: make this config/constant?
+    fn max_retries(&self) -> usize {
+        2
+    }
```

Comment:
We have #321 to make num retries and timeout into config params for the 1.10 release, could follow smtg similar?

Reply:
yes- I want to wait until we can pipe configs through to processors more easily

Reply:
Not related to this PR, but in the future we should prob have a system that allows each processor to have a "context" that can be accessed/passed around from where ever... The context would just contain  processor specific stuff like Arcs of channel txs/rxs, in-memory cache, config, etc

Been thinking of ways to do this but think the best is just to have a trait in SDK 🤣

Reply:
100000%%%%